helm: Improve the Helm chart documentation.

bmcustodio opened the pull request:
`helm-docs` has an issue that causes some of our fields to have wrong documentation associated with them (e.g. containing documentation belonging to other, commented-out fields). This PR attempts to fix them by using a fork of `helm-docs` containing the [proposed fix](https://github.com/norwoodj/helm-docs/pull/99) and adjusting what needs to be adjusted to have a minimally readable and correct reference.

**Goals:**

* Propose that we use the fork of `helm-docs` until https://github.com/norwoodj/helm-docs/pull/99 is merged.
* That all Helm chart fields present in the tables have a readable, correct and meaningful description.
  * I can definitely use a second pair of eyes on this one 🙂
* Include a comment on `helm-values.rst` indicating it is automatically generated.
* For the same reason, add `helm-values.rst` to `.gitattributes`.

**Non-goals:**

* That all Helm chart fields have an actual description.
* To fix all inconsistency issues (e.g. some descriptions end with a `.` and some don't, some descriptions use backticks for correct rendering of code, some don't).
* To make the tables extra-readable and actually eye-pleasant.

I think that these non-goals require a big effort and are only ever useful if the introduction of the fork is consensual, so I think we could cover them as part of a subsequent PR. Thoughts?

```release-note
Improve the Helm chart documentation.
```

qmonnet commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -586,8 +588,6 @@ hubble:
       # ref: https://github.com/prometheus-operator/prometheus-operator/blob/master/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml)
       enabled: false
 
-  metricsServer: ""
```
Can you please comment on why this value is removed?

bmcustodio replied:
I `git grep`ped for it and it was unused in the chart.

qmonnet commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -114,6 +114,7 @@ affinity:
           - cilium
       topologyKey: kubernetes.io/hostname
 
+# -- The priority class to use for cilium-agent
```
I understand that the purpose is not to harmonise punctuation and other formatting details in the file, but still it would be nice if the documentation comments added in the PR were consistent. Would you mind adding (or removing) the periods to all the comments you add, please?

bmcustodio replied:
Sure makes total sense.

qmonnet commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -768,16 +755,10 @@ hubble:
-      #       requests:
-      #         cpu: 100m
-      #         memory: 64Mi
+      # -- Resource requests and limits for the 'hubble-ui' container of the 'hubble-ui' deployment.
```
How does helm-docs work, would there be a way to keep the examples in the comments in this file in such a way that it's not extracted when generating the RST page? I see that most comments use `# --` as a delimiter, but given that the example is currently parsed and added to the RST I don't suppose simply keeping it with a `#` (instead of `# --`), above the actual `# --` description, would be enough to have it skipped by `helm-docs`?

In case I'm not clear, thinking of something like this, but I'm not sure if it would work:
```yaml
      # [skipped by helm-docs] example for 'resources':
      #     resources:
      #       limits:
      #         cpu: 1000m
      #         memory: 1024M
      #       requests:
      #         cpu: 100m
      #         memory: 64Mi
      # -- Resource requests and limits for the 'hubble-ui' container of the 'hubble-ui' deployment.
```

bmcustodio replied:
It works indeed ( with the forked version, I mean 🙂 ). I've added them back.

qmonnet reviewed (changes requested):
Looks great overall, thanks!

Please find a few comments and questions below.

Two additional remarks (that don't necessarily have to be addressed in the current PR):

- I think that `Documentation/helm-values.rst` could be added to `.gitattributes` so that GitHub automatically folds its diff by default for reviews, given that it is a generated file?
- Do you think we should have a commented header in `Documentation/helm-values.rst` that states that the file is generated, just like for `Documentation/cmdref/cilium-agent.md` for example? That might limit the risk of having contributors opening PRs to edit it directly?

qmonnet reviewed (commented):
The changes look good to me.

Please update the description of the PR to document your latest change, and please also document it in your commit log (I would recommend against empty commit logs in general).

Not blocking here, but my preference would even be to have several commits in this PR (one for the `helm-docs` image change, one for the doc updates, one for .gitattribute and “auto-generated” header, for example). I find it easier for understanding logical changes.

bmcustodio committed:
Use a fork of 'helm-docs'.

'helm-docs' has a bug which causes it to include comments belonging to
previously-appearing but commented-out fields. A fix has been proposed
in https://github.com/norwoodj/helm-docs/pull/99, but hasn't been
reviewed yet. While said PR doesn't get merged it's preferable to switch
to a fork containing the fix so we can have a proper description for our
Helm chart fields.

Signed-off-by: Bruno Miguel Custódio <brunomcustodio@gmail.com>

bmcustodio committed:
Add 'helm-values.rst' to '.gitattributes'.

Adding so that GitHub automatically folds its diff by default for
reviews given that it is a generated file.

Signed-off-by: Bruno Miguel Custódio <brunomcustodio@gmail.com>

bmcustodio committed:
Include codegen warning in 'helm-values.rst'.

Include a comment in 'helm-values.rst' indicating that the file is
generated automatically. This will hopefully limit the risk of having
contributors opening PRs to edit it directly.

Signed-off-by: Bruno Miguel Custódio <brunomcustodio@gmail.com>

bmcustodio committed:
Add missing descriptions to 'Helm Reference'.

Some fields appear in the 'Helm Reference' page without an associated
description. This commit aims at fixing that.

Signed-off-by: Bruno Miguel Custódio <brunomcustodio@gmail.com>

qmonnet commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1469,7 +1483,7 @@ preflight:
   podAnnotations: {}
 
-  # Labels to be added to preflight pods
+  # -- Labels to be added to thee preflight pod.
```
Thou should'st consider speaking  'i  a moe recent English, haply.

bmcustodio replied:
Forsooth - thanketh thee f'r noticing t!

bmcustodio committed:
Fix typo.

Signed-off-by: Bruno Miguel Custódio <brunomcustodio@gmail.com>

qmonnet reviewed (approved):
All good, thanks a lot!

errordeveloper reviewed (approved):
Thanks for doing this @bmcustodio, LGTM!

kaworu reviewed (approved):
Thanks!

kaworu commented on Documentation/spelling_wordlist.txt:
```diff
@@ -908,6 +910,7 @@ xfrm
 xml
 xor
 xoring
+xt
```
what is `xt`?

qmonnet replied:
I think it comes from `xt_socket`, it looks like Sphinx splits words on underscores :(.

We could maybe change that behaviour by implementing a [dedicated filter](https://sphinxcontrib-spelling.readthedocs.io/en/latest/customize.html?#custom-word-filters) for the spelling plugin :thinking:

bmcustodio replied:
I agree — but maybe as a subsequent PR that takes care of all these cases (and eventually of addressing the non-goals)?

kaworu replied:
@qmonnet Thanks for clarifying!

qmonnet replied:
@bmcustodio I don't know if it's worth doing at all, I was mostly thinking out loud. But yeah, definitely for another PR if we decided to tackle this.

kaworu reviewed (commented):
Backporting this patch to v1.10 and v1.9 won't be trivial as the Helm doc stuff has moved since the releases. I think we should abandon attempting to backport to v1.8 since our Helm charts were very different and also I don't think we generated anything using Helm doc at the time.

@bmcustodio can you consider making PRs for v1.10 and v1.9 instead of the `need-backport` labels please? I think it would save some time to the backporting team.
